### PR TITLE
secp256k1: Implement direct signature verification.

### PR DESCRIPTION
**Pull request by davecgh**
**This requires PR #2057**.

This implements signature verification directly instead of calling the version in the standard library in order to ultimately be able to break the reliance on big ints it requires and thus allow for optimization specifically for the sec256k1 curve.

It currently is only marginally faster since it still relies on big ints.  However, there is already a 10% reduction in allocations which reduces pressure on the GC during IBD when there are a ton of signature verifications happening.

```
benchmark            old ns/op    new ns/op    delta
-----------------------------------------------------
BenchmarkSigVerify   224238       219603       -2.07%

benchmark            old allocs   new allocs   delta
-----------------------------------------------------
BenchmarkSigVerify   64           57           -10.94%

benchmark            old bytes    new bytes    delta
-----------------------------------------------------
BenchmarkSigVerify   3138         2818         -10.20%
```

### CODE REVIEW

File: dcrec/secp256k1/signature.go [hunk 1]
```diff
@@ -93,7 +95,80 @@ func (sig *Signature) Serialize() []byte {
...
+	//
+	// 1. Fail if R and S are not in [1, N-1]
+	// 2. e = H(m)
+	// 3. w = S^-1 mod N
+	// 4. u1 = e * w mod N
+	//    u2 = R * w mod N
+	// 5. X = u1G + u2Q
+	// 6. Fail if X is the point at infinity
+	// 7. x = X.x mod N (X.x is the x coordinate of X)
+	// 8. Verified if x == r
```
Comment (matheusd) [inline dcrec/secp256k1/signature.go]: I think this should be `x == R` (capital R instead of lower case r).

File: dcrec/secp256k1/signature.go [hunk 2]
```diff
@@ -93,7 +95,80 @@ func (sig *Signature) Serialize() []byte {
...
+	}
+
+	// Step 7.
+	//
+	// x = X.x mod N (X.x is the x coordinate of X)
+	x.Mod(x, N)
+
+	// Step 8.
+	//
+	// Verified if x == r
```
Comment (matheusd) [inline dcrec/secp256k1/signature.go]: Same as previous message

File: dcrec/secp256k1/signature.go (hunk 1 above)
Comment (davecgh) [reply dcrec/secp256k1/signature.go]: Right. It's `r` in the book, but I agree it should match the rest of the comments which I already adapted accordingly.  Updated.

File: dcrec/secp256k1/signature.go [hunk 3]
```diff
@@ -93,7 +95,80 @@ func (sig *Signature) Serialize() []byte {
...
+	// 7. x = X.x mod N (X.x is the x coordinate of X)
+	// 8. Verified if x == R
+
+	// Step 1.
+	//
+	// Fail if R and S are not in [1, N-1].
+	N := curveParams.N
+	if sig.R.Sign() <= 0 || sig.S.Sign() <= 0 ||
+		sig.R.Cmp(N) >= 0 || sig.S.Cmp(N) >= 0 {
+
```
Comment (dnldd) [inline dcrec/secp256k1/signature.go]: is the extra space here intentional?

Comment (davecgh) [reply dcrec/secp256k1/signature.go]: That's how gofmt does it.

Comment (dnldd) [reply dcrec/secp256k1/signature.go]: noted.

**Pull request merged**